Started work on conflict API

ptersilie opened the pull request:
** PR in progress **

The goal of this PR is to provide a nice API for the user (i.e. the language developer) to inspect shift/reduce and reduce/reduce conflicts and handle them manually.

At the moment the API is **very** basic, but I thought I get some early thoughts on the direction before continuing. An example usage of the API can be found in nimbleparse, where it just pretty prints the conflicts to the terminal.

My main question would be, where the user would be expected to handle these errors. For example within the calc examples, would that be in `main.rs` or `build.rs`? Or am I off track again, and we want to allow the user to handle the conflict as it happens and they then can manually change how the parser generator resolves those conflicts?

ltratt commented:
>My main question would be, where the user would be expected to handle these errors. For example within the calc examples, would that be in main.rs or build.rs?

The user can handle these wherever they want, which means whenever they have access to a statetable. At the moment, unfortunately, there's an asymmetry between `RTParserBuilder` (where the user explicitly builds statetables themselves and can thus do what you've done with nimbleparse) and `CTParserBuilder` (where the statetable is built for them and where they can't get access to the conflicts). We need to think what to do about the latter.

My initial suggestion is the following. First we add a `bool` to `CTParserBuilder` called something like `error_on_conflicts` which defaults to true. If set to true, the user loads a grammar in with conflicts, `process_file_in_src` will return an `Err` of some sort. If set to false, `process_file_in_src` will set (new) attributes `sr_conflicts` and `rr_conflicts` in `CTParserBuilder` to whatever `stable` returned, and the user can access them (or ignore them) as they wish.

>Or am I off track again, and we want to allow the user to handle the conflict as it happens and they then can manually change how the parser generator resolves those conflicts?

I don't think we need to worry about that sort of thing yet (maybe not ever!).

ptersilie commented:
> My initial suggestion is the following. First we add a bool to CTParserBuilder called something like error_on_conflicts which defaults to true. If set to true, the user loads a grammar in with conflicts, process_file_in_src will return an Err of some sort. If set to false, process_file_in_src will set (new) attributes sr_conflicts and rr_conflicts in CTParserBuilder to whatever stable returned, and the user can access them (or ignore them) as they wish.

Yes, that sounds sensible. I had something similar in mind, minus the `error_on_conflicts` bool.

ptersilie commented on cfgrammar/src/lib/yacc/grammar.rs:
```diff
@@ -555,6 +555,26 @@ where
         }
     }
 
+    /// Returns the name of a given Symbol `sym`.
```
I moved the function into the place where it's used. Rather than documenting why it panics, maybe we should just pretty print EOF as `"EOF"`?

ltratt replied:
No, we can't give it any name, because that might clash with a user defined name for another token (and then confusion will ensue!).

ptersilie replied:
That is true.

ptersilie replied:
Anyways, now that I moved this into `prod_to_string`, this can't happen, as EOF can't be a symbol within a production, right? So I can add an `unreachable!()` for that case.

ltratt replied:
Correct.

ptersilie commented on cfgrammar/src/lib/yacc/grammar.rs:
```diff
@@ -555,6 +555,26 @@ where
+    pub fn prod_to_string(&self, pidx: PIdx<StorageT>) -> String {
+        let mut sprod = String::new();
+        let ridx = self.prod_to_rule(pidx);
+        sprod.push_str(self.rule_name(ridx));
```
I added the rule name as otherwise the conflict information is not as helpful. I also checked that your above example still works in `cfgrammar`.

ltratt replied:
Perfect!

ptersilie commented on lrtable/src/lib/statetable.rs:
```diff
@@ -96,9 +96,9 @@ pub struct StateTable<StorageT> {
     tokens_len: TIdx<StorageT>,
     /// The number of reduce/reduce errors encountered.
-    pub reduce_reduce: u64,
+    pub reduce_reduce: Vec<(PIdx<StorageT>, PIdx<StorageT>)>,
```
Then it's probably not worth the hassle. The production + rule name should be enough to find the location within the grammar manually. I'll just add the state for now.

ltratt replied:
OK.

ltratt replied:
BTW, there is a `pp_core_states` function on `StateGraph` which might be useful for you?

ptersilie replied:
Yes I've seen that. Maybe we should allow users to save that to a file if they want to look deeper into a certain conflict?

ltratt replied:
I don't like the idea of saving to a file, but we might want to expose the string to the user in some way or another.

ptersilie replied:
Ok, I guess if we give them the string, they can do whatever they want with it.

ltratt replied:
Exactly.

ltratt replied:
Perhaps what we need to do is return some sort of "conflicts struct", which, amongst other things, has the stategraph?

ltratt replied:
One method called `conflicts` which returns `Option<Conflicts>` would probably be nicer.

ptersilie commented:
The last commit takes a shot at making grammar conflicts available from within `CTParserBuilder`. This API needs some more work and there's a few things I'm not super happy with, but wanted to discuss them here first.

As discussed this adds `error_on_conflicts` to `CTParserBuilder` which, if set, errors if the grammar contains any conflicts. I wanted to give the user the option here to get some useful information out of the error, instead of just `Error: Grammar contains conflicts`. So the error contains grm, sgraph, and stable, and functions to pretty print the results. The debug print of the error is something like `Error: CTConflictsError(1 Shift/Reduce, 5 Reduce/Reduce)`, though we may want to pretty print the actual conflicts along with the graph here. Suggestions?

If `error_on_conflicts` isn't set, then the `CTParserBuilder` receives references to grm, sgraph, and stable, which the user can then manually query for conflicts and do whatever they want with them, e.g.

```
    let mut ctb = CTParserBuilder::<u8>::new()
        .error_on_conflicts(false)
        .action_kind(ActionKind::CustomAction);
    let lex_rule_ids_map = ctb
        .process_file_in_src("calc.y")?;
    panic!("{}", ctb.pp_conflicts());
```

would print out

```
Found shift/reduce conflicts:
   State 5: Shift("d") / Reduce(E: "c")
Found reduce/reduce conflicts:
   State 10: Reduce(A: "b") / Reduce(B: "a" "b")
```

The change in ctbuilder.rs:358 is also not ideal as we need to clone a string here to be able to move `grm` into `CTParserBuilder`. The alternative would be to wrap most of this function into a block so `actiontype` goes out of focus before we set `self.grm` below.

ltratt commented on lrtable/src/lib/statetable.rs:
```diff
@@ -428,14 +464,9 @@ where
-    pub fn rr_conflicts(&self) -> &Vec<(PIdx<StorageT>, PIdx<StorageT>, StIdx)> {
-        &self.reduce_reduce
+    /// Return a struct containing all conflicts or None if there aren't any.
+    pub fn conflicts(&self) -> &Option<Conflicts<StorageT>> {
```
It might be better to make this `Option<&Conflicts<...>>` to give yourself some flexibility in the future. You'd need to change the body to `self.conflicts.as_ref()`.

ltratt commented on nimbleparse/src/main.rs:
```diff
@@ -162,26 +162,26 @@ fn main() {
-                grm.pp_prod(*pidx));
+    if let Some(c) = stable.conflicts() {
+        if c.sr_len() > 0 {
+            println!("Found shift/reduce conflicts:");
```
I'm not sure the "found" bit is useful. Maybe just `"Shift/reduce conflicts:"?

ltratt commented on nimbleparse/src/main.rs:
```diff
@@ -162,26 +162,26 @@ fn main() {
-                grm.pp_prod(*r_pidx));
+        // Report reduce/reduce conflicts
+        if c.rr_len() > 0 {
+            println!("Found reduce/reduce conflicts:");
```
Same "found" comment as above.

ltratt commented on cfgrammar/src/lib/yacc/grammar.rs:
```diff
@@ -362,8 +362,13 @@ where
-    pub fn actiontype(&self) -> &Option<String> {
-        &self.actiontype
+    pub fn actiontype(&self) -> Option<&String> {
+        if let Some(ref a) = self.actiontype {
```
I think this can be `self.actiontype.as_ref()`.

ltratt commented on lrpar/examples/calc_actions/build.rs:
```diff
@@ -44,6 +44,7 @@ fn main() -> Result<(), Box<std::error::Error>> {
     // enough to fit all IDs in) as well as the input file (which must end in ".y" for lrpar, and
     // ".l" for lrlex).
     let lex_rule_ids_map = CTParserBuilder::<u8>::new()
+        .error_on_conflicts(false)
```
There are no conflicts here (I think!), so we don't want to set this to `false` do we?

ptersilie replied:
Yes, I thought I'll leave it in as an example, but I'm happy to remove it.

ltratt replied:
I want to encourage people not to do `error_on_conflicts(false)` unless they know they have conflicts (e.g. the Lua grammar), so I think we should remove this.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -73,14 +74,58 @@ pub enum ActionKind {
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let conflicts = self.stable.conflicts().unwrap();
+        write!(f, "{}", conflicts.pp(&self.grm)) //
```
Stray `//`.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -73,14 +74,58 @@ pub enum ActionKind {
+    u32: AsPrimitive<StorageT>
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let conflicts = self.stable.conflicts().unwrap();
```
This `unwrap` seems a bit dodgy: if there are no conflicts it'll cause a panic I think?

ptersilie replied:
`CTConflictsError` only gets created if there are actual any errors, so `stable.conflicts` can not be `None` here.

ltratt replied:
Aha!

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -73,14 +74,58 @@ pub enum ActionKind {
+    }
+}
+
+impl<StorageT> fmt::Debug for CTConflictsError<StorageT>
```
Can we not `#[derive(Debug)]` something that's good enough for debug support?

ptersilie replied:
Then we would also have to add `#[derive(Debug)]` to `YaccGrammar` and `StateTable`, and I'm not sure what the output would then be.

ltratt replied:
We have to implement `Debug` because of `Error` I guess?

ptersilie replied:
I'm not sure what you mean. How I see it we have 3 options:
1) We use `#[derive(Debug)]`, which means we also have to add to `YaccGrammar` and `StateTable`, and when this error is returned it prints out the entire grammar and statetable. Which IMO is a bit too much information.

2) We leave this is at is, printing out something like `CTConflictsError(1 Shift/Reduce, 5 Reduce/Reduce)` as a debug message.

3) We update this to print out the stategraph + the actual errors encountered, resulting in something like:
```
0:  [^ -> . Expr, {'$'}]
    Term -> 2
    Factor -> 4
    Expr -> 3
    'INT' -> 5
    'LBRACK' -> 1
1:  [Factor -> 'LBRACK' . Expr 'RBRACK', {'PLUS', 'MUL', 'RBRACK', '$'}]
    Expr -> 6
    'LBRACK' -> 1
    'INT' -> 5
    Factor -> 4
    Term -> 2
...

Shift/Reduce conflicts:
   State 2: Shift("PLUS") / Reduce(Term: "Term")
Reduce/Reduce conflicts:
   State 2: Reduce(Expr: "Term") / Reduce(Term: "Term")
   State 2: Reduce(Expr: "Term") / Reduce(Term: "Term")
   State 11: Reduce(Term: "Factor" "MUL" "Term") / Reduce(Term: "Term")
   State 11: Reduce(Term: "Factor" "MUL" "Term") / Reduce(Term: "Term")
   State 11: Reduce(Term: "Factor" "MUL" "Term") / Reduce(Term: "Term")

```

I'm probably tending to the last one, since then when the user sets `error_on_conflicts(true)`, they immediately get all the info they need to debug this error. Otherwise they have to first set `error_on_conflicts(false)` and then retrieve the `ConflictsStruct` from the result, and then pretty print it using a `panic!`.

ltratt replied:
I guess what I'm asking is: do we need to implement `Debug` at all?

ptersilie replied:
It seems it's needed by `Error`:
```
impl<StorageT> Error for CTConflictsError<StorageT>                                                                                                                                     
    |                ^^^^^ `ctbuilder::CTConflictsError<StorageT>` cannot be formatted using `{:?}` 
```

ltratt replied:
Right, OK, that was my guess. Then I'm more-or-less OK with this as-is. Should it be `CTConflictsError{{{} Shift/Reduce, {} Reduce/Reduce}}` to make it look more like other debug output?

ptersilie replied:
Can do. What's your thoughts on my suggestion above to output the graph and more detailed conflicts instead, so the user doesn't need to jump through so many hoops to get to this information.

ltratt replied:
I think `Debug` output should me minimal, so I'd prefer not to output the graph etc.

ptersilie replied:
Yes, I agree `Debug` should be minimal. I've moved the pretty printed stuff into the `Display` method, though I'm not sure if this is the correct place for it either.

ltratt replied:
I must admit, I generally make `Debug` and `Display` the same; and if someone wants the full pretty-printed output I'd make them call `pp`. But this is something over which reasonable people differ.

ptersilie replied:
Okay, looks go with your way then to be consistent.

ptersilie replied:
If they want a prettyprinted version for the error, they can just call `error.cs.pp()`.

ltratt replied:
Works for me.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -150,6 +197,69 @@ where
+    }
+
+    pub fn sgraph(&self) -> Option<&StateGraph<StorageT>> {
+        if let Some((_, sg, _)) = &self.error_data {
```
`self.error_data.as_ref()[1]` or similar might work here.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -150,6 +197,69 @@ where
+    }
+
+    /// Returns a pretty-printed version of the graph.
+    pub fn pp_sgraph(&self) -> String {
```
Can't someone do `sgraph().pp_core_states(&grm)`? I'm not sure we need this function.

ptersilie replied:
They can, but then they need to retrieve `grm` first from the `Conflicts` struct as it's needed as an argument. I thought this was more user friendly.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -150,6 +197,69 @@ where
+    }
+
+    /// Returns a pretty-printed version of the conflicts.
+    pub fn pp_conflicts(&self) -> String {
```
Similarly, this seems redundant.

ltratt commented:
A few high-level comments:

* `as_ref` is a wonderful function on `Option` :)
* We need tests for this stuff.
* I'm not convinced that we want so many methods on `CTBuilder` exposing stuff to the outside: the fewer we have, the lower the chance of API breakage in the future IMHO.

With regards to that last point, I think I'd prefer to only expose a `conflicts()` function with a signature along the lines of:

```rust
fn conflicts(self) -> ConflictsStruct
```

so that it consumes the `CTBuilder` (otherwise what happens if you process two files? Do the old conflicts get thrown away?). Then the `ConflictsStruct` would, minimally, have functions like `pp_sgraph` (but probably not `sgraph`).

Alternatively we could even have the `ConflictsStruct` returned from `process_file`, along the lines of:

```rust
    pub fn process_file<P, Q>(
        &self,
        inp: P,
        outd: Q
    ) -> Result<(HashMap<String, StorageT>, Option<ConflictsStruct>), Box<Error>>
    where
        P: AsRef<Path>,
        Q: AsRef<Path>
```

making it clear that the conflicts only relate to one file.

ptersilie commented on lrtable/src/lib/statetable.rs:
```diff
@@ -465,8 +491,13 @@ where
-    pub fn conflicts(&self) -> &Option<Conflicts<StorageT>> {
-        &self.conflicts
+    pub fn conflicts(&self) -> Option<&Conflicts<StorageT>> {
+        if let Some(c) = &self.conflicts {
```
I didn't know about `as_ref` but it's exactly what I needed in this case and I was thinking to myself that there must be a better way.

ptersilie commented:
> as_ref is a wonderful function on Option :)
I didn't know about it, but it's definitely what I wanted.

> I'm not convinced that we want so many methods on CTBuilder exposing stuff to the outside: the fewer we have, the lower the chance of API breakage in the future IMHO.
Agreed, I also think the fewer methods the better, but wasn't sure what needed to be exposed.

> Alternatively we could even have the ConflictsStruct returned from process_file, along the lines of
I think I like this solution the most, as then we won't have functions that return `None` or empty strings, if there weren't any errors.

ptersilie commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -238,12 +355,12 @@ where
         let mod_name = inp.as_ref().file_stem().unwrap().to_str().unwrap();
         let actiontype = match grm.actiontype() {
-            Some(t) => t,
+            Some(t) => t.clone(), // We need to clone here so we can move `grm` at the end of the function
```
Shall I leave the comment in then, so we know to remove this once this is the case?

ltratt replied:
You could maybe change it to something like "Probably unneeded once NLL is in stable"?

ptersilie commented:
See if that new commit makes things better.

ltratt commented on lrpar/examples/calc_actions/build.rs:
```diff
@@ -43,8 +43,7 @@ fn main() -> Result<(), Box<std::error::Error>> {
     // ".l" for lrlex).
-    let lex_rule_ids_map = CTParserBuilder::<u8>::new()
-        .error_on_conflicts(false)
+    let (lex_rule_ids_map, _cs) = CTParserBuilder::<u8>::new()
```
Just `_` is enough.

ltratt commented on lrpar/examples/calc_parsetree/build.rs:
```diff
@@ -43,7 +43,7 @@ fn main() -> Result<(), Box<std::error::Error>> {
     // enough to fit all IDs in) as well as the input file (which must end in ".y" for lrpar, and
     // ".l" for lrlex).
-    let lex_rule_ids_map = CTParserBuilder::<u8>::new().process_file_in_src("calc.y")?;
+    let (lex_rule_ids_map, _cs) = CTParserBuilder::<u8>::new().process_file_in_src("calc.y")?;
```
`_`

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -78,21 +78,40 @@ pub enum ActionKind {
     pub stable: StateTable<StorageT>
 }
 
+impl<StorageT> ConflictsStruct<StorageT>
```
In retrospect, this is a weird name. Shouldn't it just be `Conflicts`?

ptersilie replied:
Yes this was kind of a placeholder, I forgot to ask what we should name this. `Conflicts` is already the name of the `Conflicts` struct within `StateTable`. Which is why I wasn't sure how to name this.

ltratt replied:
That's another file and we don't import that struct in here I think, so reusing the name isn't a total disaster.

ptersilie replied:
Won't that be confusing from a documentation perspective. We'd have a struct `Conflicts`, which has a `StateTable`, which has a `Conflicts`. But the two `Conflicts` are not the same.

ptersilie replied:
Maybe I name the one in StateTable `GrammarConflicts`

ltratt replied:
They'll never appear on the same page...

I'm not saying it's ideal, but we can't add `Struct` as a suffix and I can't think of a better name...

ptersilie replied:
See my above suggestion to rename the `StateTable` one to `GrammarConflicts`.

ltratt replied:
They're all grammar conflicts in one way...

If you really want to separate the names, maybe make `StateTable`'s struct be `STConflicts`?

ptersilie replied:
Alternatively we can name the CTParserBuilder one `CTConflicts`?

ltratt commented:
Some minor comments, but I think we're heading in the right direction.

ptersilie commented:
I pushed a commit that addresses the comments, and one that adds a `Conflicts` test to `StateTable`. I'll look into adding `CTParserBuilder` tests for conflicts tomorrow.

ltratt commented on lrtable/src/lib/statetable.rs:
```diff
@@ -947,6 +947,44 @@ mod test {
+            )
+        );
+        let mut iterrr = conflicts.rr_conflicts();
+        assert_eq!(
```
Would it be easier to say `assert_eq!(conflicts.rr_conflicts().next().unwrap(), ...);`?

ptersilie commented:
I added tests for `CTParserBuilder`. Creating those tests made me realise that it'sa bit tedious to retrieve the data from `CTConflictsError` since it gets returned as a `Box<Error>`. Do we need to use a boxed type here? If so, do you have an idea how we can make it easier for the user to get the data out of the error?

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -618,3 +618,67 @@ pub fn _reconstitute<'a, StorageT: Deserialize<'a> + Hash + PrimInt + Unsigned>(
+        let temp = TempDir::default();
+        let mut file_path = PathBuf::from(temp.as_ref());
+        file_path.push("grm.y");
+        let mut f = File::create(file_path.clone()).unwrap();
```
I think you can do `File::create(&file_path)` since the argument is `AsRef`.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -618,3 +618,67 @@ pub fn _reconstitute<'a, StorageT: Deserialize<'a> + Hash + PrimInt + Unsigned>(
+B : 'a' | C;
+C : 'a';".as_bytes());
+
+        match CTParserBuilder::<u8>::new()
```
Once you merge in master you can (and probably will have to) remove the `<u8>` annotation here.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -618,3 +618,67 @@ pub fn _reconstitute<'a, StorageT: Deserialize<'a> + Hash + PrimInt + Unsigned>(
+            .action_kind(ActionKind::GenericParseTree)
+            .process_file_in_src(file_path.to_str().unwrap()) {
+            Ok((_, conflicts)) => {
+                assert_eq!(conflicts.as_ref().unwrap().stable.conflicts().unwrap().sr_len(), 1);
```
Addressing https://github.com/softdevteam/grmtools/pull/55#pullrequestreview-181277687 (assuming it's possible) will remove at least one `as_ref`.

ptersilie replied:
I'm not sure what you mean. Your link is related to statetable, here we don't have statetable but `process_in_src` returns an `Option<Conflicts>` which is moved (we can't return a reference, since the value will be out of scope once we return from the function).

ltratt replied:
Mea culpa!

However, this API is borderline unusable: `as_ref().unwrap().stable.conflicts().unwrap().sr_len()` is almost impossible to remember. Can we do *something* to make things easier? One thing might be to not expose `stable`, but instead expose things like `sr_len`? But that might cause us the same duplication-of-methods problem we had earlier in the PR...

ptersilie replied:
I can give the `CTConflictsError` struct some functions, at the very least `fn conflicts() -> Conflicts`.

ltratt replied:
That would certainly help!

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -618,3 +618,67 @@ pub fn _reconstitute<'a, StorageT: Deserialize<'a> + Hash + PrimInt + Unsigned>(
+        match CTParserBuilder::<u8>::new()
+            .action_kind(ActionKind::GenericParseTree)
+            .process_file_in_src(file_path.to_str().unwrap()) {
+            Ok((_, _)) => {
```
`Ok(_)` should be OK here.

ltratt commented:
You only need to return `Box<Error>` if you can return multiple different error types. In this case, since you can only return conflicts, I doubt you need the `Box`.

ptersilie commented:
Shall we get rid of the `Box<Error>` then?

ltratt commented:
If the complexity isn't needed... off with its head!

ptersilie commented:
Hmm, looks like `Box` is needed: `he size for values of type `(dyn std::error::Error + 'static)` cannot be known at compilation time `

ltratt commented:
Right, so when I looked, `process_file_in_src` can currently return multiple different types of errors. But I don't think anyone in their right mind will do:

```rust
        match CTParserBuilder::<u8>::new()
            .action_kind(ActionKind::GenericParseTree)
            .process_file_in_src(file_path.to_str().unwrap()) {
            Ok(_) => {
                panic!("Expected error")
            }
            Err(e) => {
                let cs = e.downcast_ref::<CTConflictsError<u8>>();
                assert_eq!(cs.unwrap().cs.stable.conflicts().unwrap().sr_len(), 1);
                assert_eq!(cs.unwrap().cs.stable.conflicts().unwrap().rr_len(), 1);
            }
        }
```

so we don't have to worry too much. Anyone with half a brain will do `error_on_conflicts(false)` and access the conflicts via the `Ok` case.

ptersilie commented:
Right. I pushed some more commits addressing your comments (935d10d) and merging in master and applying the necessary changes (1f35d6f).

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -97,6 +98,10 @@ where
         s
     }
+
+    pub fn conflicts(&self) -> &STConflicts<StorageT> {
```
Should this be `stconflicts`?

ltratt replied:
If it's confusing, it might be worth trying a different configuration of methods to find something better. I generally find my second attempt at an API is better than the first. And I've recently sometimes had to have three attempts to find something good...

ptersilie replied:
Okay, I've tried to improve it a little bit.

ltratt commented:
Dumb question: what's in `Conflicts` and what's in `ConflictsData`? Maybe there's a way to squeeze them down into one struct (e.g. perhaps there are fields like `grm: Option<&YaccGrammar>` in one of the structs?

Warning: I'm not saying this is a good idea, but I'm just trying to think of alternatives.

ptersilie commented:
`ConflictsData` just groups the graph, statetable, and grammar into one struct to return it in `process_in_src`. I guess we could also just return a tuple, but I'm not sure that makes it better. The advantage of `ConflictsData` is that it gives us some functions like `pp`.

ltratt commented:
I know what the two structs contain :) My point was more that the names don't mean anything to users and that's why I was wondering if we could combine them e.g. somethinglike:

```rust
struct Conflicts {
    reduce_reduce: Vec<(PIdx<StorageT>, PIdx<StorageT>, StIdx)>,
    shift_reduce: Vec<(TIdx<StorageT>, PIdx<StorageT>, StIdx)>
    grm: Option<YaccGrammar<StorageT>>,
    sgraph: Option<StateGraph<StorageT>>,
    stable: Option<StateTable<StorageT>>
}
```

BTW, in general we probably don't want `pub` elements on structs: in general, we should provide a method to access things (it gives us a bit more flexibility in the future).

ptersilie commented:
The problem with that is, that `Conflicts` lives on `StateTable`, so having `StateTable` within the `Conflicts` struct wouldn't make sense.

ltratt commented:
In that case, `stable` would be `None`? I'm not saying that that struct is necessarily a good idea, but can we do *something* to improve things?

ptersilie commented:
Alright, I've tried to simplify the API a bit. See if this goes in the right direction!

ltratt commented:
Can you show me an example of the new API in use? It's too hard for me to guess just from the diff.

ptersilie commented:
The tests (which also have been updated in the commit) should show it's used.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -691,16 +670,17 @@ C : 'a';"
-                assert_eq!(conflicts.as_ref().unwrap().conflicts().sr_len(), 1);
-                assert_eq!(conflicts.as_ref().unwrap().conflicts().rr_len(), 1);
+            .action_kind(ActionKind::GenericParseTree);
+        let _ = ct.process_file_in_src(file_path.to_str().unwrap());
```
`let _ = ` can be removed.

ptersilie replied:
Not without creating a warning tho.

ltratt replied:
You can fix it in nicer ways than this ;)

ltratt commented:
I can only see the `CTParserBuilder` API being tested?

ptersilie commented:
What other API do you mean?

ltratt commented:
Mea culpa! I meant: how does it look from `StateTable`?

ptersilie commented:
The `StateTable` API hasn't changed since the last time. Though I did find another `as_ref` opportunity that I just fixed and pushed.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -691,16 +670,17 @@ C : 'a';"
+        let _ = ct.process_file_in_src(file_path.to_str().unwrap());
+
+        match ct.conflicts() {
+            Some((_, _, _, conflicts)) => {
```
The `_`s are grm, sgraph, stable respectively?

ptersilie replied:
Yes, I was thinking about putting `_grm`, etc, but last time I did this you told me `_` is enough.

ltratt replied:
I didn't mean you should change them: I just wanted to check.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -256,6 +235,13 @@ where
     }
 
+    pub fn conflicts(&self) -> Option<(&YaccGrammar<StorageT>, &StateGraph<StorageT>, &StateTable<StorageT>, &Conflicts<StorageT>)> {
+        if let Some((ref grm, ref sgraph, ref stable)) = self.cdata {
```
Can you simplify this to `if let Some((grm, sgraph, stable)) = &self.cdata`? I'm not sure.

ptersilie replied:
Looks like I can! :-)

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -256,6 +235,13 @@ where
 
+    pub fn conflicts(&self) -> Option<(&YaccGrammar<StorageT>, &StateGraph<StorageT>, &StateTable<StorageT>, &Conflicts<StorageT>)> {
+        if let Some((ref grm, ref sgraph, ref stable)) = self.cdata {
+            return Some((&grm, &sgraph, &stable, &stable.conflicts().unwrap()));
```
I'm not sure you need the `&` on `grm` and `sgraph`.

ptersilie replied:
And you are right again! :-)

ptersilie commented:
Addressed some more comments. Hope it's getting there.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -173,6 +232,27 @@ where
+        self
+    }
+
+    pub fn conflicts(
```
I'm inclined to think that this API should be explicitly marked as unstable until we're sure it's in the final form we want. There isn't an off-the-shelf way to do this, but we could add an `unstable` feature to the crate (as suggested by https://github.com/rust-lang/rfcs/issues/1491#issuecomment-345455136). Thoughts?

ptersilie replied:
I have come to the opinion that this may be too confusing, and I rather just leave a comment in the documentation that this function is unstable.

ltratt replied:
Fair enough.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -565,3 +656,70 @@ pub fn _reconstitute<'a, StorageT: Deserialize<'a> + Hash + PrimInt + Unsigned>(
+
+#[cfg(test)]
+mod test {
+
```
Extraneous blank line.

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -79,14 +79,71 @@ pub enum ActionKind {
-    actionkind: ActionKind
+    actionkind: ActionKind,
+    error_on_conflicts: bool,
+    cdata: Option<(
```
`cdata` is kind of a weird name: it certainly needs to be explicitly documented what it's for, and pointed out that it's not part of the cache. You'll need to update that comment above too slightly.

ptersilie replied:
Do you have an idea for a better name? Maybe just `data`?

ltratt replied:
In a later comment, I sort-of suggest renaming this to `conflicts` (for other reasons).

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -475,6 +564,8 @@ where
         let mut f = File::create(outp_rs)?;
         f.write_all(outs.as_bytes())?;
+
+        self.cdata = Some((grm, sgraph, stable));
```
Maybe it would make more sense to have something like:

```rust
struct CTParserBuilder {
 ...
 conflicts: Option<(YaccGrammar, StateGraph, StateTable, Conflicts)>
```

and this line would be:

```rust
if let Some(c) = stable.conflicts() {
    self.conflicts = Some((grm, sgraph, stable, c));
}
```

and then `fn conflicts` would just  be `self.conflicts`.

ptersilie replied:
Since `c` is a reference, I would have to add lifetimes, which isn't really worth it for a private variable that can only be accessed via `fn conflicts` anyways. So I just renamed `cdata` to `conflicts`.

ltratt replied:
Fair enough.

ltratt commented on lrtable/src/lib/statetable.rs:
```diff
@@ -307,6 +369,14 @@ where
         let gotos_sv = SparseVec::<usize>::from(&gotos, 0, usize::from(grm.rules_len()));
 
+        let mut conflicts = None;
+        if reduce_reduce.len() > 0 || shift_reduce.len() > 0 {
```
You want to use `is_empty` otherwise Clippy will complain...

ltratt commented on lrtable/src/lib/statetable.rs:
```diff
@@ -478,21 +552,21 @@ fn resolve_shift_reduce<StorageT: 'static + Hash + PrimInt + Unsigned>(
-where
+    stidx: StIdx, // State we want to shift to
+    shift_reduce: &mut Vec<(TIdx<StorageT>, PIdx<StorageT>, StIdx)>,
+    state: StIdx // State in which the conflict occured
```
Ouch, this is a terrible name given `stidx` also exists! We have to come up with something that's not so much of a duplicate. Maybe `shift_stidx` and `conflict_stidx` or something like that?

ptersilie replied:
I changed it. But now I'm wondering if this is a similar problem and needs addressing: https://github.com/softdevteam/grmtools/blob/f8e5a1d9d8e58faf1a05dd370064ebb653548498/lrtable/src/lib/statetable.rs#L216

ltratt replied:
That's less bad, because `stidx` is `STIdx` and `state` is an entirely different type.

ptersilie commented:
See if this fixes things.

ltratt commented on lrtable/src/lib/statetable.rs:
```diff
@@ -370,7 +370,7 @@ where
 
         let mut conflicts = None;
-        if reduce_reduce.len() > 0 || shift_reduce.len() > 0 {
+        if reduce_reduce.is_empty() || shift_reduce.is_empty() {
```
Should be `!is_empty` right?

ltratt replied:
Ah, this was fixed in a later commit.

ptersilie replied:
Yes, see my `whoops` commit. ;-)

ltratt commented on lrtable/src/lib/statetable.rs:
```diff
@@ -554,7 +554,7 @@ fn resolve_shift_reduce<StorageT: 'static + Hash + PrimInt + Unsigned>(
     stidx: StIdx, // State we want to shift to
     shift_reduce: &mut Vec<(TIdx<StorageT>, PIdx<StorageT>, StIdx)>,
-    state: StIdx // State in which the conflict occured
+    shift_stidx: StIdx // State in which the conflict occured
```
This variable name seems odd, given the comment on `stidx`. Can we come up with something better?

ptersilie replied:
I was just following your suggestion, though I do agree this can be somewhat confusing. I guess your other suggestion `conflict_stidx` it is then.

ltratt commented on lrtable/src/lib/statetable.rs:
```diff
@@ -554,7 +554,7 @@ fn resolve_shift_reduce<StorageT: 'static + Hash + PrimInt + Unsigned>(
     stidx: StIdx, // State we want to shift to
     shift_reduce: &mut Vec<(TIdx<StorageT>, PIdx<StorageT>, StIdx)>,
-    shift_stidx: StIdx // State in which the conflict occured
+    conflicts_stidx: StIdx // State in which the conflict occured
```
Is it `conflicts_idx` plural or just `conflict_idx`?

ptersilie replied:
Yes, should probably be singular. But I think it should still be `conflict_stidx` as it is the id of the state the conflict occurs in.

ltratt replied:
`conflict_stidx` works for me.

ptersilie commented:
Renamed `conflicts_stidx`. This should be all now I think.

ltratt commented:
Now that we've hit 150 comments, I think it's time to squash :)

ptersilie commented:
I think so too. I'm getting quite a few conflicts when I squash everything down, so this is gonna take me a bit.

ptersilie commented:
Rebase and squashed. When the tests pass, I think we are ready to go!

ltratt commented on cfgrammar/src/lib/yacc/grammar.rs:
```diff
@@ -557,6 +557,28 @@ where
+        for sym in self.prod(pidx) {
+            let s = match sym {
+                Symbol::Token(tidx) => {
+                    if *tidx == self.eof_token_idx() {
```
The `if` statement is unnecessary: just `self.token_name(*tidx).unwrap()` is needed (it'll `panic` if it hits `eof_token_idx`).

ltratt commented on lrpar/src/lib/ctbuilder.rs:
```diff
@@ -173,6 +233,29 @@ where
+        self
+    }
+
+    /// Returns the grammar conflicts found during generation of the parser, or `None` if there
```
Should this be something like "If the grammar had conflicts, return a tuple which allows users to see what they were; or `None` if there were no conflicts"?

ltratt commented on lrtable/src/lib/statetable.rs:
```diff
@@ -424,6 +493,11 @@ where
         }
     }
+
+    /// Return a struct containing all conflicts or None if there aren't any.
```
`None` needs backticking.

ltratt commented on nimbleparse/src/main.rs:
```diff
@@ -159,6 +159,11 @@ fn main() {
     };
 
+    if let Some(c) = stable.conflicts() {
+        println!("Stategraph:\n{}\n", sgraph.pp_core_states(&grm));
```
Should we say something like "Conflicts found: {} shift/reduce {} reduce/reduce\n  {}" here? Otherwise the user might not realise what the stategraph is for?

ptersilie replied:
Hmm, I think the next line already says something like
```
Shift/Reduce conflicts:
   State: 1: Shift(2) / Reduce(E: T)
```

So we'd mention "Shift/Reduce conflicts" twice. Are you still fine with that?

ptersilie replied:
Or maybe we just swap the lines, and put the conflicts before the graph?

ltratt replied:
Swapping the lines works for me!

ltratt commented:
Doing a final read through I spotted a few minor things. Won't take more than a minute or two to fix.

Once that's done and checked, we should expand the main commit's message: it would be nice to have a brief example in the commit of the API in use (for `StateTable` and `CTParserBuilder`), because it's a bit hard to find it in the diff.

ptersilie commented:
Addressed your comments. See if that works for you.

ltratt commented:
Works for me! Please squash (and also please edit the commit message along the lines suggested a bit earlier). Almost there!

ptersilie commented:
Squashed and also edited the commit message. See what you think!
